### PR TITLE
Add category hierarchy

### PR DESCRIPTION
**Pull request by okkez**
connect to https://github.com/bm-sms/daimon-news-misc/issues/86

### CODE REVIEW

File: app/models/category.rb [hunk 1]
```diff
@@ -2,8 +2,70 @@ class Category < ActiveRecord::Base
...
+  acts_as_list scope: [:ancestry], column: :order
+
   validates :slug, format: /\A\w+\z/, uniqueness: {scope: :site_id}
-  validates :order, numericality: :only_integer
+  validates :order, numericality: {only_integer: true, allow_nil: true}
 
   scope :ordered, -> { order(:order) }
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{table_name} AS c ON c.#{ancestry_column} = CAST(#{table_name}.id AS text) OR c.#{ancestry_column} = #{table_name}.#{ancestry_column} || '/' || #{table_name}.id").group("#{table_name}.id").having("COUNT(c.id) = 0").order("#{table_name}.ancestry NULLS FIRST").order(:order) }
```
Comment (okkez) [inline app/models/category.rb]: I don't use input from the outside, so SQL injection may not be occurred.

File: app/assets/javascripts/editor/form.js.es6 [hunk 2]
```diff
@@ -1,7 +1,9 @@
-window.setupPostForm = () => {
-  function applySelect2($element, options) {
+$(() => {
+  function applySelect2(selector, options) {
```
Comment (tricknotes) [inline app/assets/javascripts/editor/form.js.es6]: It seems `options` is currently unnecessary.
When we want to pass some options, `$element.data()` is available.

File: app/controllers/editor/categories_controller.rb [hunk 3]
```diff
@@ -43,6 +43,24 @@ def destroy
     redirect_to editor_categories_url, notice: "カテゴリが削除されました"
   end
 
+  # move to left
+  def move_higher
```
Comment (tricknotes) [inline app/controllers/editor/categories_controller.rb]: Can you some actions except CRUD as other controller?
ref: http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/

File: app/models/category.rb [hunk 4]
```diff
@@ -2,8 +2,67 @@ class Category < ActiveRecord::Base
...
+
   validates :slug, format: /\A\w+\z/, uniqueness: {scope: :site_id}
-  validates :order, numericality: :only_integer
+  validates :order, numericality: {only_integer: true, allow_nil: true}
 
   scope :ordered, -> { order(:order) }
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{table_name} AS c ON c.#{ancestry_column} = CAST(#{table_name}.id AS text) OR c.#{ancestry_column} = #{table_name}.#{ancestry_column} || '/' || #{table_name}.id").group("#{table_name}.id").having("COUNT(c.id) = 0").order("#{table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
+  scope :parental, -> { includes(:posts).where("posts.id" => nil).order("#{table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
```
Comment (tricknotes) [inline app/models/category.rb]: Can you use `quoted_table_name` instead of `table_name` ?
The former is more safe as SQL string.

File: app/views/editor/categories/index.html.haml [hunk 5]
```diff
@@ -4,24 +4,4 @@
...
-    - @categories.each do |category|
-      %tr
-        %td= category.name
-        %td= category.description
-        %td= category.slug
-        %td= category.order
-        %td= link_to 'Show', editor_category_url(category)
-        %td= link_to 'Edit', edit_editor_category_url(category)
-        %td= link_to 'Destroy', editor_category_url(category), method: :delete, data: { confirm: 'Are you sure?' }
+= render partial: 'table', locals: {categories: @categories}
```
Comment (tricknotes) [inline app/views/editor/categories/index.html.haml]:
``` haml
= render 'table', categories: @categories
```

File: app/views/editor/categories/show.html.haml [hunk 6]
```diff
@@ -7,9 +7,8 @@
 %p
   %b Slug:
   = @category.slug
-%p
-  %b Order
-  = @category.order
+
+= render partial: 'table', locals: {categories: @category.children.ordered}
```
Comment (tricknotes) [inline app/views/editor/categories/show.html.haml]:
``` haml
render 'table', categories: @category.children.ordered
```

File: app/controllers/editor/categories_controller.rb (hunk 3 above)
Comment (okkez) [inline app/controllers/editor/categories_controller.rb]: Do you mean that we should create `Editor::Categories::MoveHigherController#update` or something?

File: app/assets/javascripts/editor/form.js.es6 (hunk 2 above)
Comment (okkez) [inline app/assets/javascripts/editor/form.js.es6]: I use `options` here.
https://github.com/bm-sms/daimon-news/pull/564/files#diff-bf289ed1c9275ec90f6157adf147e824R28

File: app/controllers/editor/categories_controller.rb (hunk 3 above)
Comment (tricknotes) [inline app/controllers/editor/categories_controller.rb]: It is close but not correct.
- 👍  Create some controller `Editor::Categories::XXXController#update`
- 👎  The controller name is not noun

The controller name should be a noun.
So the name will be `PositionController`, `OrderController` or something.

File: app/assets/javascripts/editor/form.js.es6 (hunk 2 above)
Comment (tricknotes) [inline app/assets/javascripts/editor/form.js.es6]: Hmm, that's true.
Does `$element.find('select')` have no `data-` attributes?

File: app/controllers/editor/categories_controller.rb (hunk 3 above)
Comment (tricknotes) [inline app/controllers/editor/categories_controller.rb]: `higher` ? `left` ?

File: app/controllers/editor/categories_controller.rb [hunk 7]
```diff
@@ -43,6 +43,24 @@ def destroy
...
+    category = categories.find(params[:id])
+    target_category = category.higher_items.last
+    category.move_to_left_of(target_category) if target_category
```
Comment (tricknotes) [inline app/controllers/editor/categories_controller.rb]: When `target_category` is `nil`?
If it is bad case, the error should be raised?

File: app/controllers/editor/categories_controller.rb [hunk 8]
```diff
@@ -43,6 +43,24 @@ def destroy
...
+
+    redirect_to editor_categories_url, notice: "カテゴリが移動されました"
```
Comment (tricknotes) [inline app/controllers/editor/categories_controller.rb]: I think it is kindness to embed `category.name` into this message.

File: app/assets/javascripts/editor/form.js.es6 (hunk 2 above)
Comment (okkez) [inline app/assets/javascripts/editor/form.js.es6]: I can add `data-` attribute in `_credit_fields.html.haml` :yum:

File: app/controllers/editor/categories_controller.rb (hunk 3 above)
Comment (okkez) [inline app/controllers/editor/categories_controller.rb]: - higher is acts_as_list word
- left is ancestry word

File: test/models/category_test.rb [hunk 9]
```diff
@@ -1,7 +1,49 @@
 require "test_helper"
 
 class CategoryTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  def setup
+    DatabaseCleaner.start
+    site = create(:site)
+    @categories = create_list(:category, 5, site: site)
```
Comment (tricknotes) [inline test/models/category_test.rb]: When the some category names are the same as other one, the test cases of category's order are silently broken :<

File: test/models/category_test.rb [hunk 10]
```diff
@@ -1,7 +1,49 @@
 require "test_helper"
 
 class CategoryTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  def setup
+    DatabaseCleaner.start
```
Comment (tricknotes) [inline test/models/category_test.rb]: Why is this needed?

Comment (okkez) [inline test/models/category_test.rb]: When remove `DatabaseCleaner.start`, append records for each test.

File: app/views/editor/categories/_form.html.haml [hunk 11]
```diff
@@ -2,7 +2,8 @@
   = f.text_field :name
   = f.text_area :description, class: "markdown-editor", data: {upload_path: editor_images_url}
   = f.text_field :slug, help: "URLの一部に使うこのカテゴリを表すわかりやすい英単語を指定します"
-  = f.number_field :order, help: "ナビゲーションバーに表示する順序を指定します"
+  - help_text = "親カテゴリを指定します。指定しない場合はトップレベルのカテゴリになります"
+  = f.collection_select :parent_id, current_site.categories.parental, :id, :full_name, {include_blank: true, help: help_text}, {class: "select2", "data-select2-allow-clear": true}
```
Comment (tricknotes) [inline app/views/editor/categories/_form.html.haml]: Except `@category` from `current_site.categories.parental` ?

File: test/models/category_test.rb (hunk 9 above)
Comment (okkez) [inline test/models/category_test.rb]: `factory :category` uses sequence for category name.
So all categories have different name each other.

On the other hand, we may need test that some categories have same name.

Comment (tricknotes) [inline test/models/category_test.rb]: > `factory :category` uses sequence for category name.

This is the current behavior and `Category` has no validation for name.
To keep this test stable, I think that to specify category name here is better way.

File: app/views/editor/categories/_form.html.haml (hunk 11 above)
Comment (okkez) [inline app/views/editor/categories/_form.html.haml]: Oh! Fixed :bow:

**Comment by okkez** [comment]
I've fixed against all comments and pushed changes.

File: app/assets/javascripts/editor/form.js.es6 [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+      return Number($(el).val());
+    });
+
+    orders.push(0); // Minimum value
+
+    let maxOrder = Math.max(...orders);
+    $element.find('.credit-order-value').val(maxOrder + 1);
+  }
+
+  applySelect2('select.select2', {});
```
Comment (tricknotes) [inline app/assets/javascripts/editor/form.js.es6]: The second argument is not assigned to variable in this function.

File: app/models/category.rb [hunk 13]
```diff
@@ -2,8 +2,78 @@ class Category < ActiveRecord::Base
...
+
+  def move_to(reference_instance, direction:)
+    case direction.to_sym
+    when :child
+      move_to_child_of(reference_instance)
+    when :left
+      move_to_left_of(reference_instance)
+    when :right
+      move_to_right_of(reference_instance)
+    end
```
Comment (tricknotes) [inline app/models/category.rb:89]: How do you think about `else` direction?

File: app/validators/post_category_validator.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+class PostCategoryValidator < ActiveModel::Validator
+  def validate(record)
+    if record.category.has_children?
+      record.errors[:category_id] << "子カテゴリを持つカテゴリ「#{record.category.full_name}」に記事を登録することはできません。"
```
Comment (tricknotes) [inline app/validators/post_category_validator.rb:4]: How about using i18n for error message?
Such as https://github.com/bm-sms/nursepress/pull/272/commits/6c2a818594f35deff5f2a1f5b676f443db5c6634.

File: app/views/categories/_categories.html.haml [hunk 15]
```diff
@@ -0,0 +1,18 @@
+%ul.category-summary
+  - categories.each do |category|
+    - if category.has_children?
+      - count = category.children.inject(0) do |memo, child|
```
Comment (tricknotes) [inline app/views/categories/_categories.html.haml]: Can you use [`Enumerable#sum`](http://api.rubyonrails.org/classes/Enumerable.html#method-i-sum) to simplify calculation?

``` ruby
category.children.map {|child| child.posts.published.count }.sum
```

File: config/routes.rb [hunk 16]
```diff
@@ -17,7 +17,9 @@
 
     resources :fixed_pages
     resources :links
-    resources :categories
+    resources :categories do
+      resources :orders, module: :categories, only: %i(update), param: :direction, constraints: {direction: /left|right/}
```
Comment (tricknotes) [inline config/routes.rb]: This route generates the following URL.

```
PUT /editor/categories/:category_id/orders/:direction
```

I think this is not RESTFul because `:direction` is not a resource.

Instead, how about the following URL?

```
POST /editor/categories/:category_id/order?move_to=:direction
```

File: app/views/categories/_categories.html.haml (hunk 15 above)
Comment (okkez) [inline app/views/categories/_categories.html.haml]: This is the expression on the inside of deep loop.
So I want to reduce loop here.

File: app/validators/post_category_validator.rb (hunk 14 above)
Comment (okkez) [inline app/validators/post_category_validator.rb:4]: In this project we don't use i18n for all messages.
If we don't need Japanese message here, I will translate to English.

File: app/models/category.rb [hunk 17]
```diff
@@ -2,8 +2,80 @@ class Category < ActiveRecord::Base
...
+      move_to_left_of(siblings.find(array_of_ids.second))
+    else
+      move_to_right_of(siblings.find(array_of_ids[array_of_ids.index(id.to_s) - 1]))
+    end
+  end
+
+  def move_to_child_of(reference_instance)
+    transaction do
+      remove_from_list
+      update_attributes!(parent: reference_instance)
```
Comment (tricknotes) [inline app/models/category.rb]: Can you use `update!` (it is the alias of `update_attributes!`) for consistency?

```
$ git grep "update\!" | wc
       9      26     658
$ git grep "update_attributes\!" | wc
       0       0       0
```

File: app/models/category.rb [hunk 18]
```diff
@@ -2,8 +2,80 @@ class Category < ActiveRecord::Base
   belongs_to :site
   has_many :posts
 
+  has_ancestry
+  acts_as_list scope: [:ancestry], column: :order
+
   validates :slug, format: /\A\w+\z/, uniqueness: {scope: :site_id}
-  validates :order, numericality: :only_integer
+  validates :order, numericality: {only_integer: true, allow_nil: true}
```
Comment (tricknotes) [inline app/models/category.rb:9]: When the order becomes `nil`?

Comment (okkez) [inline app/models/category.rb:9]: When `remove_from_list`, this is `acts_as_list`'s method.

File: app/views/categories/_categories.html.haml (hunk 15 above)
Comment (mtsmfm) [inline app/views/categories/_categories.html.haml]: > So I want to reduce loop here.

Hmm, it seems we can't reduce loop at any rate.

In this case, `sum` is more explanatory than `inject` I think.

---

FYI: `sum` accepts block:

``` ruby
count = category.children.sum {|child| child.posts.published.count }
```

File: app/controllers/editor/categories/orders_controller.rb [hunk 19]
```diff
@@ -0,0 +1,26 @@
+class Editor::Categories::OrdersController < Editor::ApplicationController
+  def create
```
Comment (mtsmfm) [inline app/controllers/editor/categories/orders_controller.rb]: `update` ?

File: app/views/categories/_categories.html.haml (hunk 15 above)
Comment (okkez) [inline app/views/categories/_categories.html.haml]: https://github.com/rails/rails/blob/a678c47f6d31977e2a91cc198f115c803f0fc8e1/activesupport/lib/active_support/core_ext/enumerable.rb#L20

`sum` accepts block, but calls `map` and `inject`.

In my case, calls `inject` only and loop children at once.

In tricknotes case,

``` ruby
category.children.map {|child| child.posts.published.count }.sum
```

This invokes `map` and `inject` (in `sum`), loop children and array (has same size with children).
loop twice.

In mtsmfm case,

``` ruby
count = category.children.sum {|child| child.posts.published.count }
```

This invokes `map` in `sum` and call `sum` recursively and call `inject`.
After all, loop twice.

File: app/views/categories/_categories.html.haml [hunk 20]
```diff
@@ -0,0 +1,18 @@
+%ul.category-summary
+  - categories.each do |category|
+    - if category.has_children?
+      - count = category.children.sum do |memo, child|
```
Comment (kou) [inline app/views/categories/_categories.html.haml]: The first block argument `memo` is needed?

**Comment by okkez** [comment]
Rebased

File: app/assets/javascripts/editor/form.js.es6 [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  function setupCreditOrder($element) {
+    let orders = $('.credit-order-value').map((_i, el) => {
+      return Number($(el).val());
+    });
+
+    orders.push(0); // Minimum value
+
+    let maxOrder = Math.max(...orders);
```
Comment (tricknotes) [inline app/assets/javascripts/editor/form.js.es6:18]: We can simplify 🙇 💦 

``` js
Math.max(0, ...orders);
```

File: app/controllers/editor/categories/orders_controller.rb [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    redirect_to editor_categories_url, notice: message
+  end
+
+  private
+
+  def categories
+    current_site.categories
+  end
+
+  def category_params
```
Comment (tricknotes) [inline app/controllers/editor/categories/orders_controller.rb]: Maybe is this unused?

File: app/views/editor/categories/_category.html.haml [hunk 23]
```diff
@@ -0,0 +1,11 @@
+%tr{class: "depth#{category.depth}"}
+  %td= category.name
+  %td= category.description
+  %td= category.slug
+  %td= link_to 'Show', editor_category_url(category)
+  %td= link_to 'Edit', edit_editor_category_url(category)
+  %td= link_to 'Destroy', editor_category_url(category), method: :delete, data: { confirm: 'Are you sure?' }
+  %td= link_to_unless category.first?, '▲', editor_category_order_url(category_id: category, move_to: "left", target: category.higher_item&.id), method: :patch
```
Comment (tricknotes) [inline app/views/editor/categories/_category.html.haml]: Should this HTTP method be the `PUT`?

File: app/models/category.rb [hunk 24]
```diff
@@ -2,8 +2,80 @@ class Category < ActiveRecord::Base
...
+        update!(order: lower_item_order)
+      else
+        add_to_list_bottom
+        save!
+      end
+    end
+  end
+
+  def move_to(reference_instance, direction:)
+    case direction.to_sym
```
Comment (tricknotes) [inline app/models/category.rb:80]: Are there any reason using `to_sym`?
I think this should be a `String` because `direction` comes from user input.

File: app/models/category.rb (hunk 18 above)
Comment (tricknotes) [inline app/models/category.rb:9]: confirmed

File: app/models/category.rb (hunk 24 above)
Comment (okkez) [inline app/models/category.rb:80]: Just normalized.

File: db/seeds/development/posts.seeds.rb [hunk 25]
```diff
@@ -70,20 +72,53 @@
...
-      EOS
+      body: paginate_body,
       thumbnail: Rails.root.join("db/data/thumbnail.jpg").open,
       published_at: Time.current,
       category: site1.categories.find_by!(slug: "category2"),
     )
   end
+
+  25.times do |i|
+    i += site1.posts.maximum(:id)
```
Comment (tricknotes) [inline db/seeds/development/posts.seeds.rb]: We can use `public_id` instead `id`.
The latter is not published anywhere.

File: test/models/category_test.rb [hunk 26]
```diff
@@ -1,7 +1,44 @@
...
+    assert_equal(1, @categories[3].order)
+  end
+
+  def test_move_to_left_of
+    @categories[4].move_to_left_of(@categories[3])
+    expected = @categories.map(&:name).values_at(0, 1, 2, 4, 3)
+    assert_equal(expected, Category.ordered.pluck(:name))
+  end
+
+  def test_move_to_left_of_2
```
Comment (tricknotes) [inline test/models/category_test.rb]: What is the difference between 1 and 2?
I'm happy if it is described at test case.

File: app/views/editor/categories/_category.html.haml (hunk 23 above)
Comment (okkez) [inline app/views/editor/categories/_category.html.haml]: I'm not sure.

File: app/assets/javascripts/editor/form.js.es6 (hunk 21 above)
Comment (okkez) [inline app/assets/javascripts/editor/form.js.es6:18]: Could you fix this after merge this PR?

File: app/models/category.rb (hunk 24 above)
Comment (tricknotes) [inline app/models/category.rb]: nod

File: app/views/editor/categories/_category.html.haml (hunk 23 above)
Comment (tricknotes) [inline app/views/editor/categories/_category.html.haml]: > PUT to a URL creates/replaces the resource in its entirety at the client defined URL.
> PATCH to a URL updates part of the resource at that client defined URL.

ref: http://stackoverflow.com/questions/630453/put-vs-post-in-rest

In this case, these request parameters are the part of resource or itself?

File: app/assets/javascripts/editor/form.js.es6 (hunk 21 above)
Comment (tricknotes) [inline app/assets/javascripts/editor/form.js.es6:18]: nod

File: app/views/editor/categories/_category.html.haml (hunk 23 above)
Comment (okkez) [inline app/views/editor/categories/_category.html.haml]: I got it :clap:

**Comment by okkez** [comment]
Rebased

File: app/models/category.rb [hunk 27]
```diff
@@ -2,8 +2,90 @@ class Category < ActiveRecord::Base
...
+  acts_as_list scope: [:ancestry], column: :order
+
   validates :slug, format: /\A\w+\z/, uniqueness: {scope: :site_id}
-  validates :order, numericality: :only_integer
+  validates :order, numericality: {only_integer: true, allow_nil: true}
 
   scope :ordered, -> { order(:order) }
+  # For PostgreSQL, doesn't work on MySQL
+  # See https://github.com/stefankroes/ancestry/pull/238
+  scope :leaves, -> { joins("LEFT JOIN #{quoted_table_name} AS c ON c.#{ancestry_column} = CAST(#{quoted_table_name}.id AS text) OR c.#{ancestry_column} = #{quoted_table_name}.#{ancestry_column} || '/' || #{quoted_table_name}.id").group("#{quoted_table_name}.id").having("COUNT(c.id) = 0").order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
```
Comment (tricknotes) [inline app/models/category.rb]: Can you specify `order` only once?

``` diff
- order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order)
+ order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST", :order)
```

File: app/models/category.rb [hunk 28]
```diff
@@ -2,8 +2,90 @@ class Category < ActiveRecord::Base
...
+  scope :parental, ->(category = nil) { includes(:posts).where("posts.id" => nil).where.not(id: category&.id).order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order) }
```
Comment (tricknotes) [inline app/models/category.rb]: `order("#{quoted_table_name}.#{ancestry_column} NULLS FIRST").order(:order)` is used twice.
Can you extract it as scope?

File: test/models/category_test.rb [hunk 29]
```diff
@@ -1,7 +1,44 @@
 require "test_helper"
 
 class CategoryTest < ActiveSupport::TestCase
-  # test "the truth" do
-  #   assert true
-  # end
+  def setup
+    site = create(:site)
+    @categories = create_list(:category, 5, site: site)
```
Comment (tricknotes) [inline test/models/category_test.rb]: Can you specify each category to individual name?
Currently `Category#name` depends on data initialization.

File: test/test_helper.rb [hunk 30]
```diff
@@ -13,4 +13,5 @@
 class ActiveSupport::TestCase
   # Add more helper methods to be used by all tests here...
   include FactoryGirl::Syntax::Methods
+  self.use_transactional_fixtures = false
```
Comment (tricknotes) [inline test/test_helper.rb:16]: 👍

File: test/models/category_test.rb [hunk 31]
```diff
@@ -1,7 +1,44 @@
...
+    assert_equal(2, @categories[3].order)
+    @categories[3].move_to_child_of(@categories[4])
+    assert_equal("#{@categories[0].id}/#{@categories[4].id}", @categories[3].ancestry)
+    assert_equal(1, @categories[3].order)
+  end
+
+  def test_move_to_left_of
+    @categories[4].move_to_left_of(@categories[3])
+    expected = @categories.map(&:name).values_at(0, 1, 2, 4, 3)
+    assert_equal(expected, Category.ordered.pluck(:name))
```
Comment (tricknotes) [inline test/models/category_test.rb:29]: This test case depends on `Category.ordered`.
I think this test case is a unit test, so it shouldn't extra logic.

File: test/integration/editor_test.rb [hunk 32]
```diff
@@ -307,25 +306,122 @@ class EditorTest < ActionDispatch::IntegrationTest
...
+      end
+      within(:row, "Python") do
+        assert_false(find("td", text: "▲").has_css?("a"))
+        assert_true(find("td", text: "▼").has_css?("a"))
+      end
+
+      click_on("New Category")
+
+      fill_in("Name",        with: "JRuby")
+      fill_in("Description", with: "JRuby is a one of Ruby implementation.")
```
Comment (tricknotes) [inline test/integration/editor_test.rb]: "JRuby is an implementation of Ruby" ?

File: app/models/category.rb (hunk 28 above)
Comment (okkez) [inline app/models/category.rb]: Good catch. But I could not find out good name for this ...
I want to extract this when I write same code 3 times.

File: test/models/category_test.rb (hunk 31 above)
Comment (okkez) [inline test/models/category_test.rb:29]: I want to test ordered categories after move categories.

---

入れ替えたあとのカテゴリの並びをテストしたい、という説明をしてご納得いただいた。

Comment (tricknotes) [inline test/models/category_test.rb:29]: 👍

**Comment by mtsmfm** [comment]
:shipit:

**Comment by mtsmfm** [comment]
I released migration.
Could you rebase this pr?

**Comment by mtsmfm** [comment]
:shipit:
